storage/innobase/include/ut0counter.h

```diff
@@ -99,7 +100,19 @@ struct MY_ALIGNED(CACHE_LINE_SIZE) ib_counter_t
 		/* Check that we aren't writing outside our defined bounds. */
 		for (size_t i = 0; i < UT_ARR_SIZE(m_counter); i += n) {
 			for (size_t j = 1; j < n - 1; ++j) {
-				ut_ad(m_counter[i + j] == 0);
+				if (sizeof(Type) == 8) {
+					ut_ad(my_atomic_load64_explicit(
+						  reinterpret_cast<int64*>(
+						      &m_counter[i + j]),
+						  MY_MEMORY_ORDER_RELAXED)
+					      == 0);
+				} else if (sizeof(Type) == 4) {
+					ut_ad(my_atomic_load32_explicit(
+						  reinterpret_cast<int32*>(
+						      &m_counter[i + j]),
+						  MY_MEMORY_ORDER_RELAXED)
+					      == 0);
+				}
```

Comment:
Just thoughts...

Destructor must not be called concurrently with `add()`, so we can omit atomic operation here. Which is what we normally did before to indicate that this is non-concurrent access.

However with C++11 mixed (atomic and non-atomic) access to variables seems to be not allowed (although it seems to be possible with C++20).

I leave it up to you to decide if you want to follow original rule or keep it as is.

Reply:
I've thought about that too. I will remove dtor because ASAN is able check such stuff.